src/librustc_typeck/check/demand.rs

```diff
@@ -175,6 +178,61 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
+    // In cases such as:
+    //
+    // ```
+    // struct Foo {}
+    // struct Bar {}
+    //
+    // fn hey(pouet: Vec<Bar>){}
+    //
+    // fn main() {
+    //     let foo = vec![Foo {}];
+    //     hey(foo); // shouldn't suggest ".to_vec()"
+    // }
+    // ```
+    //
+    // Or in here:
+    //
+    // ```
+    // fn foo(b: &[u16]) {}
+    //
+    // fn main() {
+    //     let a: Vec<u8> = Vec::new();
+    //     foo(&a); // shouldn't suggest ".as_slice()"
+    // }
+    // ```
+    //
+    // The primitive type is considered the same even though it shouldn't. It fixes it a bit.
+    fn is_really_a_match(&self, ty_: Ty<'tcx>, expected: Ty<'tcx>) -> bool {
+        match (self.get_ty(expected), self.get_ty(ty_)) {
+            (Some(expected), Some(ty_)) => expected == ty_,
+            _ => true,
+        }
+    }
+
+    // For now it only checks slices, arrays and Vecs.
+    fn get_ty(&self, ty_: Ty<'tcx>) -> Option<Ty<'tcx>> {
+        match ty_.sty {
+            ty::TypeVariants::TyRawPtr(t) | ty::TypeVariants::TyRef(_, t) => self.get_ty(t.ty),
+            ty::TypeVariants::TyArray(ty_, _) | ty::TypeVariants::TySlice(ty_) => {
+                Some(ty_)
+            }
+            ty::TypeVariants::TyAdt(_, subs) => {
+                if ty_.to_string().split("::").last().unwrap_or("").starts_with("Vec<") {
```

Comment:
Yep... I needed to check if it was a `Vec` but if you have a better to check it, I'm all for it!